Common scripts and types generator

mutantcornholio opened the pull request:
Fixes paritytech/opstooling#125, #13

I couldn't find schemas that should be common from the start, so I added
an example one, which we will remove after adding real schemas.

This package adds `opstooling-scripts` .bin target, which is defined in
src/cli

`generate-types` is the first available command there.

mutantcornholio committed:
Common scripts and types generator

Fixes paritytech/opstooling#125, #13

I couldn't find schemas that should be common from the start, so I added
an example one, which we will remove after adding real schemas.

This package adds `opstooling-scripts` .bin target, which is defined in
src/cli

`generate-types` is the first available command there.

mordamax commented:
@mutantcornholio how do we instrument the projects with the generator, may be you're planning to integrate it somewhere?
I'm thinking may be makes sense to add some `readme.md` for instrumentization instructions?

mutantcornholio commented:
> @mutantcornholio how do we instrument the projects with the generator, may be you're planning to integrate it somewhere? 

I was thinking on `vanity-service`, as I have a fresh example there. There are also couple of valid examples in `cla-bot-2021` and `gitspiegel`.

> I'm thinking may be makes sense to add some `readme.md` for instrumentization instructions?

Sure thing.

joao-paulo-parity commented on src/cli/utils.ts:
```diff
@@ -0,0 +1,23 @@
+
+  cb(ctx).catch((err) => {
+    program.error(
+      err instanceof Error ? err.stack ?? err.message : String(err),
```
There's a function for this already called `displayError` from `./error.ts`

joao-paulo-parity commented on src/cli/commands/generate-types.ts:
```diff
@@ -0,0 +1,36 @@
+      if (options.prettier) {
+        ctx.logger.info("Applying prettier to generated files")
+
+        execSync(`prettier --write "${options.out}"`)
```
While I don't see quoting becoming a problem here, `spawnSync` (https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options) can be used instead as that accepts `string[]` for the arguments, so that way it's not necessary to deal with quoting at all.

mutantcornholio replied:
There's an important difference in `spawn` vs. `exec` behaviour. `exec` throws an error on non-zero exit, which I think is the correct behaviour in this case: 

```
Welcome to Node.js v16.13.1.
Type ".help" for more information.
> child_process.execSync("exit 1")
Uncaught Error: Command failed: exit 1
    at checkExecSyncError (node:child_process:826:11)
    at Object.execSync (node:child_process:900:15) {
  status: 1,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 70101,
  stdout: <Buffer >,
  stderr: <Buffer >
}
> child_process.spawnSync("exit", ["1"], {shell: true})
{
  status: 1,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 70228,
  stdout: <Buffer >,
  stderr: <Buffer >
}
```

joao-paulo-parity replied:
> exec throws an error on non-zero exit, which I think is the correct behaviour in this case

Yes, definitely.

Without the `{shell:true}` option, the result of `spawnSync` contains an `.error` or otherwise a `result.status != 0`

```console
> # .error is set because "exit" is not found in $PATH
> node -e 'console.log(require("child_process").spawnSync("exit", ["1"]).error)'
Error: spawnSync exit ENOENT
    at Object.spawnSync (node:internal/child_process:1111:20)
    at Object.spawnSync (node:child_process:814:24)
    at [eval]:1:38
    at Script.runInThisContext (node:vm:129:12)
    at Object.runInThisContext (node:vm:305:38)
    at node:internal/process/execution:76:19
    at [eval]-wrapper:6:22
    at evalScript (node:internal/process/execution:75:60)
    at node:internal/main/eval_string:27:3 {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawnSync exit',
  path: 'exit',
  spawnargs: [ '1' ]
}
> # .status is set to nonzero because "cat" is found in $PATH, but the command fails
node -e 'console.log(require("child_process").spawnSync("cat", ["???"]))'
{
  status: 1,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 63 61 74 3a 20 27 3f 3f 3f 27 3a 20 4e 6f 20 73 75 63 68 20 66 69 6c 65 20 6f 72 20 64 69 72 65 63 74 6f 72 79 0a>
  ],
  pid: 645698,
  stdout: <Buffer >,
  stderr: <Buffer 63 61 74 3a 20 27 3f 3f 3f 27 3a 20 4e 6f 20 73 75 63 68 20 66 69 6c 65 20 6f 72 20 64 69 72 65 63 74 6f 72 79 0a>
}
```

So it could be

```js
const result = require("child_process").spawnSync("cat", ["???"])

if (result.error) {
  throw result.error
}

if (result.status ?? 0 !== 0) {
  console.error(result.stderr.toString())
  process.exit(result.status || 1)
}
```

(which would make sense to put in a helper function)

With the `{shell:true}` option you can check only for `result.status != 0` instead.

I couldn't quite intuit why the `shell: true` option would be needed. Is it for the prettier configuration for that given project to be picked up? If so, the `cwd` option of `spawnSync` would be useful for this.

While it's hard to see how `execSync` could go wrong, `spawnSync` is overall safer in case of malformed input - which I'm not expecting since I imagine this script would be called from a preset command, so either option seems fine.

mutantcornholio replied:
> With the `{shell:true}` option you can check `result.status != 0` instead.

I know, but `execSync` is the one call that does it all in this case. You also stated that there could not be any problems with quoting, why do we need to make things more complicated than they are?

> I couldn't quite intuit why the `shell: true` option would be needed. Is it for the prettier configuration for that given project to be picked up? If so, the `cwd` option of `spawnSync` would be useful for this.

> Without the `{shell:true}` option, the result of `spawnSync` contains an `.error`

It contains an `ENOENT` error, because, there's no `exit` binary found, as without shell, PATH resolution is not performed.  
It's not the error because `exit 1` exited with non-zero code, `exit` binary in that case was never executed.  

So, I'm using shell here for the same purpose, to resolve `prettier` binary from PATH.

joao-paulo-parity replied:
> I know, but execSync is the one call that does it all in this case. You also stated that there could not be any problems with quoting, why do we need to make things more complicated than they are?

We're not _foreseeing_ any problems from how we _expect_ the command to be used. It doesn't mean it's a more reliable option. As I mentioned it leaves the door open for malformed input to have potentially damaging effects and it's not as safe since input isn't being sanitized for the shell.

> binary found, as without shell, PATH resolution is not performed.
> So, I'm using shell here for the same purpose, to resolve prettier binary from PATH.

`spawnSync` also resolves from $PATH. A quick test shows that

```console
$ mkdir tmp
$ cd tmp
$ touch foo
$ chmod +x foo
$ # note PATH="$PWD:$PATH"
$ PATH="$PWD:$PATH" node -e 'console.log(require("child_process").spawnSync("foo"))'
{
  status: 0,
  signal: null,
  output: [ null, <Buffer >, <Buffer > ],
  pid: 681459,
  stdout: <Buffer >,
  stderr: <Buffer >
}
> node -e 'console.log(require("child_process").spawnSync("foo"))'
{
  error: Error: spawnSync foo ENOENT
      at Object.spawnSync (node:internal/child_process:1111:20)
      at Object.spawnSync (node:child_process:814:24)
      at [eval]:1:38
      at Script.runInThisContext (node:vm:129:12)
      at Object.runInThisContext (node:vm:305:38)
      at node:internal/process/execution:76:19
      at [eval]-wrapper:6:22
      at evalScript (node:internal/process/execution:75:60)
      at node:internal/main/eval_string:27:3 {
    errno: -2,
    code: 'ENOENT',
    syscall: 'spawnSync foo',
    path: 'foo',
    spawnargs: []
  },
  status: null,
  signal: null,
  output: null,
  pid: 681721,
  stdout: null,
  stderr: null
}
```

In any case I'm not going to insist on convicing in terms of what's more reliable since I haven't had success with that so far. If you want to keep it like this, ok.

mutantcornholio replied:
Wait, I was wrong. All three of spawn, exec, execFile do resolve PATH

My example didn't work as `exit` is a built-in shell command.  
Unlike `true`, which can be both an executable and built-in command, there's no `/bin/exit`.  
This being said, I'll change it to `execFileSync("prettier", ["--write", options.out])`

joao-paulo-parity commented on src/cli/opstooling-scripts.sh:
```diff
@@ -0,0 +1,10 @@
+# namely tsnode and tsconfig-paths.
+# `readlink` is important to get around a symlink in node_modules/.bin,
+# so we would get actual location of the script
+script_dir="$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")"
```
The `readlink` command can fail and that's not being handled here. When that happens the command would turn into `dirname ""`, which outputs `.`. In that case `cd` would work since `cd .` always works, but the bug would go unnoticed.

A safer option would be to

```bash
set -eu -o pipefail

script_path="$(readlink -f "${BASH_SOURCE[0]}")"
script_dir="$(dirname "$script_path")"
```

With the `set -eu -o pipefail` bit it's not necessary to be checking for bad exit codes anywhere since those failures automatically exit the program; with that setup we focus on handling the expected failures in a few specific places, rather than everywhere.

```bash
# this command can fail
foo bar || :
```

All that being said, I think this could be fairly simply into a small Node.js script, which is not subject to Bash's weirdness while being more portable (less tools to depend on and is cross-platform).

```js
#!/usr/bin/env node

// fs.readLink can read the file's location from process.argv[0], but perhaps __dirname can work as well
// `cd` is not necessary because you can set the command's working directory with `spawnSync.options.cwd`

console.log(process.argv)
```

It's not a problem to put a shebang at the start, Node.js is fine with that.

mutantcornholio replied:
Yeah... both points make sense.
I'll see how js script would look like, and either go that direction or with `set -eu -o pipefail
`

mutantcornholio committed:
Codereview fixes

mutantcornholio committed:
Don't try fixing trailing whitespaces on markdown

Double trailing whitespaces are line breaks in markdown.
For some reason, this is turned off by default.
